airbyte-integrations/connectors/source-google-analytics-v4/source_google_analytics_v4/source.py

```diff
@@ -0,0 +1,423 @@
+        Returns tuple with access token and token's time-to-live
+        """
+        response_json = None
+        try:
+            response = requests.request(method="POST", url=self.token_refresh_endpoint, params=self.get_refresh_request_params())
+
+            response_json = response.json()
+            response.raise_for_status()
+        except requests.exceptions.RequestException as e:
+            if response_json and "error" in response_json:
+                raise Exception(
+                    "Error refreshing access token {}. Error: {}; Error details: {}; Exception: {}".format(
+                        response_json, response_json["error"], response_json["error_description"], e
+                    )
+                ) from e
+            raise Exception(f"Error refreshing access token: {e}") from e
+        else:
+            return response_json["access_token"], response_json["expires_in"]
+
+    def get_refresh_request_params(self) -> Mapping[str, any]:
+        """
+        Sign the JWT with RSA-256 using the private key found in service account JSON file.
+        """
+        token_lifetime = 3600  # token lifetime is 1 hour
+
+        iat = time.time()
+        exp = iat + token_lifetime
+
+        payload = {
+            "iss": self.client_email,
+            "sub": self.client_email,
+            "scope": self.scope,
+            "aud": self.token_refresh_endpoint,
+            "iat": iat,
+            "exp": exp,
+        }
+
+        headers = {"kid": self.client_id}
+
+        signed_jwt = jwt.encode(payload, self.client_secret, headers=headers, algorithm="RS256")
+
+        return {"grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer", "assertion": str(signed_jwt)}
+
+
+class SourceGoogleAnalyticsV4(AbstractSource):
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        try:
+            auth = GoogleAnalyticsOauth2Authenticator(config)
+            token = auth.get_access_token()
+            url = f"{GoogleAnalyticsV4Stream.url_type_list}"
+
+            session = requests.get(url, headers={"Authorization": "Bearer {}".format(token)})
+            session.raise_for_status()
+            return True, None
+        except requests.exceptions.RequestException as e:
+            return False, e
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        streams: List[GoogleAnalyticsV4Stream] = []
+
+        authenticator = GoogleAnalyticsOauth2Authenticator(config)
+
+        config["authenticator"] = authenticator
+        config["ga_streams"] = json.loads(pkgutil.get_data("source_google_analytics_v4", "defaults/default_reports.json")) + json.loads(
+            config["custom_reports"]
+        )
+
+        for stream in config["ga_streams"]:
+            # construct GAReadStreams sub-class for each stream
+            stream_class = type(stream["name"], (GoogleAnalyticsV4ObjectsBase,), {"cursor_field": "ga_date"})
```

Comment:
Do we guarantee that the class property `{"cursor_field": "ga_date"}` is not going to be changed in the future, and will be used by every stream that is generated?

Reply:
Yes, the only ga_date could be cursor_field for GAv4 streams